server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -162,6 +169,12 @@ protected SearchHit nextDoc(int doc) throws IOException {
                 if (context.isCancelled()) {
                     throw new TaskCancelledException("cancelled");
                 }
+                ++processedDocs;
```

Comment:
We could potentially not check the breaker when finishing collecting from a segment and rely on other parallel fetch operations running in the system to check the real memory breaker. 

If we decide to carry on as it is now and **check at the end of a segment** this has the potentially unwanted (or perhaps it is wanted?) side effect of checking once / segment even if the source is never requested.

Reply:
Made this change in https://github.com/elastic/elasticsearch/pull/121920/commits/abacdb16d580ba2eb0b11ca75546df50da420fb6 but let me know what you think @original-brownbear

Reply:
Hmm as a I pointed out above, any check at a higher rate than the TLAB region size is probably not a good idea. 
But now that you're asking :) Why do we even bother wit the per-leaf cost and reset the counts on a per-leaf basis. If we just don't reset on a new segment this problem goes away and the behavior is generally less affected by index geometry as well which is always nice?

Reply:
As discussed, moved to per fetch phase accounting as inter segment concurrency is not available in the fetch phase. 
Added minimum size of 1MiB for the buffer and no upper limit (configuring a large value could enable us to disable the feature completely)